Add ADR about Azure private link use (#743)
18F/piipan: docs/securing-internal-apis.md

- Some Azure resources can function as *both* a client *and* a server. Take for example Piipan's Orchestrator API. When it is *called* by the Query Tool, it is operating as the server. When it is *calling* the individual state APIs, it is operating as the client. In this case the Orchestrator API has two service principals: one for its system-assigned identity when operating as a client, and one for its local application object instance when it is operating as a server.
- App Service Authentication [handles all HTTP requests](https://docs.microsoft.com/en-us/azure/app-service/overview-authentication-authorization#on-windows) before they reach the application. Any requests that reach the application have already been authorized. Consequently, when an Azure Function is utilizing App Service Authentication, the application itself is configured with an [`AuthLevel` of `Anonymous`](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=csharp#configuration).

## Virtual Network Integration

> Since Azure resources are all PaaS systems, Azure uses what they call Private Links and Private Endpoints within Virtual Networks (VNet) to achieve a virtualized and entirely private setup.

Reviewer: ```suggestion
Azure Private Links, Private Endpoints, and Virtual Networks (VNets) can be used to achieve a virtualized and entirely private network for communication between Azure resources.
```
